Store new document in visit object

daun opened the pull request:
**Description**

- Store the parsed `Document` of the new page in the visit object as `visit.to.document`
- Allow reuse in plugins and consumer hooks to improve performance
- Remove the parsed document when the visit ends, is aborted, or throws an exception
- Read the containers from the visit object directly instead of duplicating them as arguments

**Questions**

- Are we rewriting the html in any plugins where we'd also need to rewrite the document?

**Checks**

- [x] The PR is submitted to the `master` branch
- [x] The code was linted before pushing (`npm run lint`)
- [x] All tests are passing (`npm run test`)
- [x] New or updated tests are included
- [ ] The documentation was updated as required

**Context**

- Closes #862

hirasso commented:
> Are we rewriting the html in any plugins where we'd also need to rewrite the document?

I can't think of any

hirasso commented on tests/unit/replaceContent.test.ts:
```diff
@@ -54,13 +59,14 @@ describe('replaceContent', () => {
 		`);
 		const warn = vi.spyOn(console, 'warn');
-		const page = mockPage(/*html*/ `
+		const html = `
```
The `/* html */` comment tells this vscode extension to correctly syntax-highlight the html template literal:

https://marketplace.visualstudio.com/items?itemName=Tobermory.es6-string-html

daun replied:
Good to know! I've added them back.

daun commented:
@hirasso Feel free to give this a spin locally! Tests are passing and I've been toying with it in the playground, but it's best to get more eyes on this as we'll be probably start relying on the new property in plugins. At least for body class and head plugin, from what I can think of.

hirasso commented:
So, I gave this a spin. It's working well! Of course, this makes it easy to break swup:

```
swup.hooks.before("content:replace", (visit: Visit) => {
  visit.to.document?.querySelector('#swup')?.remove();
});
```

So it's probably worth to mention in the docs that people should be careful when working with `visit.to.document`.

Another thing I noticed while testing this: on `page:load`, `visit.to.html` and `visit.to.document` aren't accessible, as we are only putting them in the `visit` after that hook. I imagine people would intuitively reach for the `page:load` hook to grab the document as early as possible. Is there a reasoning behind this or can we move it inside the hook?

daun commented:
@hirasso Those are good points.

> Of course, this feature makes it easy to break swup

Breaking swup has become easier and easier with replaceable hooks and a writable visit object, but I see that as a net-positive. I like tools I can bend to work the way If want to 🤠 Mentioning that in the docs is probably a good idea, but I wouldn't lose sleep over it.

> people should be careful when working with visit.to.document and remember to cloneNode(true) things if they want to extract elements from it

Your comment reminded me to double-check we're not actually removing any elements from the new document while replacing them in `replaceContent`. I'll do that tomorrow.

> on page:load, visit.to.html and visit.to.document aren't accessible, as we are only putting them in the visit after that hook

I agree — having access to the new document would be handy and intuitive in `page:load`. There is a reason it's currently implemented outside of the hook: the hook is supposed to be replaceable without duplicating its logic, e.g. by the preload plugin that returns a preload promise instead. Adding this logic to the hook handler would require the preload plugin to duplicate this logic, returning us to the previous state of duplicating logic across the core and its plugins. Having this outside of a hook ensures it always happens.

Can we think of any other way of making this available? Personally, I wouldn't know what to do with that data since the timing of `page:load` is random and there's more useful hooks later during the lifecycle, but consumers would be kind of right to expect it there.

hirasso commented:
> Breaking swup has become easier and easier with replaceable hooks and a writable visit object, but I see that as a net-positive. I like tools I can bend to work the way If want to 🤠 Mentioning that in the docs is probably a good idea, but I wouldn't lose sleep over it.

I sure love the hackability of _modern_ swup 🥳 Lets just not forget to mention this in the docs. OR not mention it at all for now 🙃

> Your comment reminded me to double-check we're not actually removing any elements from the new document while replacing them in replaceContent. I'll do that tomorrow.

Nice!

> I agree — having access to the new document would be handy and intuitive in page:load. There is a reason it's currently implemented outside of the hook: the hook is supposed to be replaceable without duplicating its logic, e.g. by the preload plugin that returns a preload promise instead. Adding this logic to the hook handler would require the preload plugin to duplicate this logic, returning us to the previous state of duplicating logic across the core and its plugins. Having this outside of a hook ensures it always happens.

> Can we think of any other way of making this available? Personally, I wouldn't know what to do with that data since the timing of page:load is random and there's more useful hooks later during the lifecycle, but consumers would be kind of right to expect it there.

Let's leave it as-is. I was probably making up a problem where none exists 😄

daun commented:
@hirasso You had the correct hunch! The current implementation was still missing `cloneNode(true)` to avoid modifying the document and inadvertently removing the incoming containers. Would have totally missed that. Fixed, and added a unit test.

hirasso commented:
Nice! Shall I re-approve then?

hirasso commented:
Does parallel plugin overwrite `replaceContent`? I'm AFK at the moment and can't check

daun commented:
Parallel Plugin only hooks before and after `content:replace`. It leaves the default handler in place so that it can handle any non-parallel containers still left in `visit.containers`.

hirasso commented:
Do we have to check other plugins as well for this issue?
